src/Tools/dotnet-monitor/Commands/CollectCommandHandler.cs

```diff
@@ -143,5 +149,55 @@ private static IHostBuilder Configure(this IHostBuilder builder, StartupAuthenti
                 }
             });
         }
+
+        private static void WatchStdinForDisconnect(IServiceProvider serviceProvider, CancellationToken cancellationToken)
+        {
+            try
+            {
+                IHostApplicationLifetime lifetime = serviceProvider.GetService<IHostApplicationLifetime>();
+                using Stream inputStream = Console.OpenStandardInput();
+                byte[] buffer = new byte[100];
+
+                async Task WatchAsync()
+                {
+                    try
+                    {
+                        while (true)
+                        {
+                            if (await inputStream.ReadAsync(buffer.AsMemory(), cancellationToken) == 0)
+                            {
+                                // input stream closed
+                                break;
+                            }
+                        }
+                    }
+                    catch (IOException)
+                    {
+                        // In case the input stream returns a OS error, we will just exit.
+                    }
+                    catch (UnauthorizedAccessException)
+                    {
+                        // In case the input stream returns ERROR_ACCESS_DENIED, we will just exit.
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        return;
+                    }
+
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        return;
+                    }
+
+                    lifetime.StopApplication();
+                }
+
+                _ = WatchAsync();
+            }
+            catch (Exception e)
+            {
+                throw new MonitoringException(string.Format(CultureInfo.InvariantCulture, Strings.ErrorMessage_UnableToWatchForDisconnect, e.Message), e);
```

Comment:
Oh, I misunderstood what the comment about `MonitoringException` was saying. Sounds like I should be writing the output to standard error?.